sql/handler.h

```diff
@@ -1405,6 +1405,23 @@ struct handlerton
    int  (*recover)(handlerton *hton, XID *xid_list, uint len);
    int  (*commit_by_xid)(handlerton *hton, XID *xid);
    int  (*rollback_by_xid)(handlerton *hton, XID *xid);
+   /*
+     recover_rollback_by_xid is optional. If set, it will be called instead
+     of rollback_by_xid when transactions should be rollback at server startup.
+     This function should do a asynchronously rollback, thus rolling back a long
+     transaction will not block server startup.
+
+     The rollback status should be persisted after this call, it should
+     continue to rollback after server restart automatically independent
+     on binlog recovery process.
+   */
+   int  (*recover_rollback_by_xid)(const XID &xid);
+   /*
+     It is called after binlog recovery has done commit/rollback of
+     all transactions. It is used together with recover_rollback_by_xid()
+     together to rollback prepared transactions asynchronously.
+   */
+   void (*signal_tc_log_recovery_done)();
    /*
```

Comment:
I think this comment could be clarified a bit.

As written, "rollback status should be persisted after this call", it sounds like recover_rollback_by_xid() should make the changes durable (ie. sync the InnoDB WAL). And it is not clear what the purpose of signal_tc_log_recovery_done() is.

But I understand from the patch that it is actually signal_tc_log_recovery_done() that is responsible for making durable the decision inside innodb to rollback all the transactions that recover_rollback_by_xid() was called on. As an optimization to not need an fsync() for every call of recover_rollback_by_xid().

It would be good to explain this in the comment. That recover_rollback_by_xid() marks that the transaction/XID should be rolled back, but does not need to persist this decision to disk yet, and does not need to complete the actual rollback immediately. And that the purpose of signal_tc_log_recovery_done() is to persist this all to disk(). This would help another storage engine author to know what to do to implement these.

Reply:
updated the comment. I found that persist the status change is not really necessary. it will be persisted by the following undo actions. If crash happens before it is persisted, it will be in prepared state and the binlog recover will rollback it again.

Reply:
updated